src/futils.c

```diff
@@ -26,30 +26,28 @@ int git_futils_mkpath2file(const char *file_path, const mode_t mode)
 
 int git_futils_mktmp(git_str *path_out, const char *filename, mode_t mode)
 {
+	const int open_flags = O_RDWR | O_CREAT | O_EXCL | O_BINARY | O_CLOEXEC;
+	unsigned int tries = TMP_MAX;
 	int fd;
-	mode_t mask;
 
-	p_umask(mask = p_umask(0));
+	while (tries--) {
+		git_str_sets(path_out, filename);
+		git_str_puts(path_out, "_git2_XXXXXX");
```

Comment:
We could help mktemp() out a bit by adding a couple of random number so the template string before passing it on.
```diff
-               git_str_sets(path_out, filename);
-               git_str_puts(path_out, "_git2_XXXXXX");
+               git_str_clear(path_out);
+               git_str_printf(path_out, "%s_git2_%04XXXXXXX", filename, rand());
```
It could be useful for systems that base the mktemp string on the PID.
As for possible DoS attacks I think that a malicious person with write access to the repository can do a bit more damage than just that.

Reply:
I wouldn't want to use `rand` - we can't be guaranteed that the caller ever invoked `srand` in a meaningful way and we wouldn't want to do it ourselves because they may be relying on an `srand` seed for some sort of reproducibility.

I appreciate your reminder that this is limited to the repository:

> a malicious person with write access to the repository can do a bit more damage than just that.

I'd been thinking about this function as a generic temporary file creation utility; since we don't use this to write into `/tmp`, I don't think that we need to be too careful about it.  I added a note on the header so that we remember its usefulness and merged it manually.

Thanks again for this!